presto-native-execution/presto_cpp/main/PeriodicTaskManager.cpp

```diff
@@ -643,4 +650,51 @@ void PeriodicTaskManager::addHttpEndpointLatencyStatsTask() {
       kHttpEndpointLatencyPeriodGlobalCounters,
       "http_endpoint_counters");
 }
+
+void PeriodicTaskManager::addWatchdogTask() {
+  addTask(
+      [this,
+       deadlockedTasks = std::vector<std::string>(),
+       opCalls = std::vector<velox::exec::Task::OpCallInfo>()]() mutable {
+        deadlockedTasks.clear();
+        opCalls.clear();
+        if (!taskManager_->getLongRunningOpCalls(
+                std::chrono::milliseconds(stuckDriverThresholdMs_),
+                stuckDriverThresholdMs_,
+                deadlockedTasks,
+                opCalls)) {
+          LOG(ERROR)
+              << "Cannot take lock on task manager, likely starving or deadlocked";
+          RECORD_METRIC_VALUE(kCounterNumTasksDeadlock, 1);
+          detachWorker();
+          return;
+        }
+        for (auto& taskId : deadlockedTasks) {
+          LOG(ERROR) << "Starving or deadlocked task: " << taskId;
+        }
+        RECORD_METRIC_VALUE(kCounterNumTasksDeadlock, deadlockedTasks.size());
+        for (auto& call : opCalls) {
+          LOG(ERROR) << "Stuck operator: tid=" << call.tid
+                     << " taskId=" << call.taskId << " opId=" << call.opId;
+        }
+        RECORD_METRIC_VALUE(kCounterNumStuckDrivers, opCalls.size());
+        if (!deadlockedTasks.empty() || !opCalls.empty()) {
+          detachWorker();
+        }
+      },
+      60'000'000, // 60 seconds
+      "Watchdog");
+}
+
+void PeriodicTaskManager::detachWorker() {
+  LOG(ERROR) << velox::process::TraceContext::statusLine();
+  if (server_ && server_->nodeState() == NodeState::kActive) {
+    // Benefit of shutting down is that the queries that aren't stuck yet will
+    // be finished.  While stopping announcement would kill them.
+    LOG(ERROR)
```

Comment:
I know we discussed internally. But shutdown state is specifically when the server gets a SIGTERM. This is breaking the semantics of the NodeState and can confuse coordinator with wrong state. Other options can be marking worker unhealthy or disable announcement. Did we observer some queries were ending successfully when worker got stuck?

Reply:
@amitkdutta 

We didn't have enough data to see observe if some queries would finish successfully as we only seen this phenomenon 5 times exactly and late enough that everything was blocked.
However, I don't see why this might not be the case.

Since INACTIVE has that special weird meaning/limitation, I actually start seeing the SHUTTING_DOWN state as a good candidate to put the worker into to isolate the whole cluster from the problem.
I think about it like this:
1. We detected a service-breaking issue.
2. Our first reflex was 'restart' and continue.
3. But then we thought: what about debugging? And what about some queries that can still finish successfully?

So putting ourselves in the SHUTTING_DOWN state and waiting for an engineer to debug looks very logical in this light.